Mathlib/LinearAlgebra/Matrix/NonsingularInverse.lean

```diff
@@ -361,6 +362,92 @@ lemma mul_right_injective_of_inv (A : Matrix m n α) (B : Matrix n m α) (h : A
+  simp_rw [←mulVec_transpose, mulVec_surjective_iff_exists_rightInverse]
+  refine ⟨fun ⟨B, hB⟩ ↦ ⟨Bᵀ, ?_⟩, fun ⟨B, hB⟩ ↦ ⟨Bᵀ, by rw [←transpose_mul, hB, transpose_one]⟩⟩
+  apply_fun Matrix.transpose using transpose_injective
+  rw [transpose_mul, transpose_transpose, hB, transpose_one]
+
+theorem vecMul_surjective_iff_isUnit {A : Matrix m m α} :
+    Function.Surjective A.vecMul ↔ IsUnit A := by
+  rw [vecMul_surjective_iff_exists_leftInverse]
+  exact ⟨fun ⟨B, hBA⟩ ↦ isUnit_of_left_inverse hBA,
+    fun h ↦ ⟨A⁻¹, @inv_mul_of_invertible _ _ _ _ _ A h.invertible ⟩⟩
+
+theorem mulVec_surjective_iff_isUnit {A : Matrix m m α} :
+    Function.Surjective A.mulVec ↔ IsUnit A := by
+  rw [←isUnit_transpose, ←vecMul_surjective_iff_isUnit]; convert Iff.rfl; rw [vecMul_transpose]
+
+theorem mulVec_injective_iff_isUnit {A : Matrix m m K} :
+    Function.Injective A.mulVec ↔ IsUnit A := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · rw [←isUnit_transpose, ←vecMul_surjective_iff_isUnit]
+    simp_rw [vecMul_transpose]
+    exact LinearMap.surjective_of_injective (f := A.mulVecLin) h
+  change Function.Injective A.mulVecLin
+  rw [←LinearMap.ker_eq_bot, LinearMap.ker_eq_bot']
+  intro c hc
+  replace h := h.invertible
+  simpa using congr_arg A⁻¹.mulVecLin hc
+
+theorem vecMul_injective_iff_isUnit {A : Matrix m m K} :
+    Function.Injective A.vecMul ↔ IsUnit A := by
+  simp_rw [←mulVec_transpose, mulVec_injective_iff_isUnit, isUnit_transpose]
+
+theorem linearIndependent_rows_iff_isUnit {A : Matrix m m K} :
+    LinearIndependent K (fun i ↦ A i) ↔ IsUnit A := by
+  rw [←transpose_transpose A, ←mulVecLin_injective_iff, mulVecLin_transpose,
+    transpose_transpose, ←vecMul_injective_iff_isUnit, coe_vecMulLinear]
+
+theorem linearIndependent_cols_iff_isUnit {A : Matrix m m K} :
+    LinearIndependent K (fun i ↦ Aᵀ i) ↔ IsUnit A := by
+  rw [←transpose_transpose A, isUnit_transpose, linearIndependent_rows_iff_isUnit,
+    transpose_transpose]
+
+theorem vecMul_surjective_of_invertible (A : Matrix m m α) [Invertible A] :
+    Function.Surjective A.vecMul :=
+  fun y ↦ ⟨A⁻¹.vecMul y, by simp⟩
+
+theorem mulVec_surjective_of_invertible (A : Matrix m m α) [Invertible A] :
+    Function.Surjective A.mulVec :=
+  fun y ↦ ⟨A⁻¹.mulVec y, by simp⟩
+
+theorem vecMul_injective_of_invertible (A : Matrix m m K) [Invertible A] :
+    Function.Injective A.vecMul :=
+  vecMul_injective_iff_isUnit.2 <| isUnit_of_invertible A
+
+theorem mulVec_injective_of_invertible (A : Matrix m m K) [Invertible A] :
+    Function.Injective A.mulVec :=
+  mulVec_injective_iff_isUnit.2 <| isUnit_of_invertible A
+
+theorem linearIndependent_rows_of_invertible (A : Matrix m m K) [Invertible A] :
+    LinearIndependent K (fun i ↦ A i) :=
+  linearIndependent_rows_iff_isUnit.2 <| isUnit_of_invertible A
+
+theorem linearIndependent_cols_of_invertible (A : Matrix m m K) [Invertible A] :
+    LinearIndependent K (fun i ↦ Aᵀ i) :=
+  linearIndependent_cols_iff_isUnit.2 <| isUnit_of_invertible A
```

Comment:
Generally `Invertible` should only be used when you are actually using the data of the `invOf` operator. Here you're not, so `isUnit` would be more appropriate.

Reply:
These lemmas are analogous to `mul_right_injective_of_invertible` etc in the section above. What makes them different?

Reply:
Ugh, you're right. Sounds like a future cleanup task for me, rather than a reasonable request on this PR.